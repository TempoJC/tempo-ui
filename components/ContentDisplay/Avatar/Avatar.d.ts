import { AvatarProps } from './types';
/**
 * If an image is used it will be shown as avatar, otherwise the initials will be used.
 */
declare const Avatar: ({ className, imageUrl, initials, size, onClick, color, shape, type, testId, }: AvatarProps) => import("react/jsx-runtime").JSX.Element;
export default Avatar;
