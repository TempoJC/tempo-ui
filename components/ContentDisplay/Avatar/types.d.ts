import { CommonProps } from '../../../types';
import { MouseEventHandler } from '../../../../node_modules/react';
export type AvatarProps = CommonProps & {
    imageUrl?: string;
    initials: string;
    onClick?: MouseEventHandler<HTMLDivElement>;
    shape?: "circle" | "square";
    size?: "small" | "medium" | "large";
    type?: "filled" | "outlined";
};
