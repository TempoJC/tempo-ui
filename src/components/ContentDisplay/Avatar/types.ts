import { CommonProps } from "@/types";
import type { MouseEventHandler } from "react";

export type AvatarProps = CommonProps & {
  imageUrl?: string;
  initials: string;
  onClick?: MouseEventHandler<HTMLDivElement>;
  shape?: "circle" | "square";
  size?: "small" | "medium" | "large";
  type?: "filled" | "outlined";
};
