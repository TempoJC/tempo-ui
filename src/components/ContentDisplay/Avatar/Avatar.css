.tempo-avatar {
  background: var(--color-token-light-background-2);
  width: var(--size7);
  height: var(--size7);
  display: flex;
  justify-content: center;
  align-items: center;
  position: relative;
  cursor: default;
  user-select: none;
  color: var(--color-token-light-content-1);
  font: var(--typography-label-l-regular-font);
  letter-spacing: var(--typography-label-l-regular-letter-spacing);
  text-transform: var(--typography-label-l-regular-text-transform);
  box-sizing: border-box;
}

.tempo-avatar--circle {
  border-radius: 100%;
}

.tempo-avatar--color-dark {
  color: var(--color-token-dark-content-1);
  background: var(--color-token-dark-background-2);
}

.tempo-avatar--large {
  font: var(--typography-title-m-font);
  text-transform: var(--typography-title-m-text-transform);
  letter-spacing: var(--typography-title-m-letter-spacing);
  height: var(--size11);
  width: var(--size11);
}

.tempo-avatar--medium {
  font: var(--typography-title-s-font);
  text-transform: var(--typography-title-s-text-transform);
  letter-spacing: var(--typography-title-s-letter-spacing);
  height: var(--size9);
  width: var(--size9);
}

.tempo-avatar--clickable {
  cursor: pointer;
}

.tempo-avatar__image {
  background-size: cover;
  background-position: center;
  width: 100%;
  height: 100%;
  border-radius: inherit;
}

.tempo-avatar--outlined {
  border: var(--tempo-border-width) solid var(--color-token-light-border-1);
  background: var(--color-token-light-background-1);
}

.tempo-avatar--color-dark.tempo-avatar--outlined {
  border: var(--tempo-border-width) solid var(--color-token-dark-border-1);
  background: var(--color-token-dark-background-1);
}
