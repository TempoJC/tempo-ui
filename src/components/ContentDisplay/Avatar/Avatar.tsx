import { useEffect, useState } from "react";
import cn from "classnames";
import { AvatarProps } from "./types";
import "./Avatar.css";

/**
 * If an image is used it will be shown as avatar, otherwise the initials will be used.
 */
const Avatar = ({
  className,
  imageUrl,
  initials,
  size = "small",
  onClick,
  color = "light",
  shape = "square",
  type = "filled",
  testId,
}: AvatarProps) => {
  const [loadError, setLoadError] = useState(false);

  useEffect(() => {
    if (!imageUrl) {
      return;
    }

    const image = new Image();
    let canceled = false;

    image.onload = () => {
      if (!canceled) {
        setLoadError(false);
      }
    };

    image.onerror = () => {
      if (!canceled) {
        setLoadError(true);
      }
    };

    image.src = imageUrl;

    return () => {
      canceled = true;
    };
  }, [imageUrl]);

  return (
    <div
      className={cn("tempo-avatar", className, {
        "tempo-avatar--clickable": Boolean(onClick),
        "tempo-avatar--large": size === "large",
        "tempo-avatar--medium": size === "medium",
        "tempo-avatar--color-dark": color === "dark",
        "tempo-avatar--outlined": type === "outlined",
        "tempo-avatar--circle": shape === "circle",
      })}
      onClick={onClick}
      data-testid={testId}
    >
      {imageUrl && !loadError ? (
        <div
          className="tempo-avatar__image"
          style={{ backgroundImage: `url(${imageUrl})` }}
          title={initials}
          data-testid={testId && `${testId}-image`}
        />
      ) : (
        <div className="tempo-avatar__initials" data-testid={testId && `${testId}-initials`}>
          {initials}
        </div>
      )}
    </div>
  );
};

export default Avatar;
