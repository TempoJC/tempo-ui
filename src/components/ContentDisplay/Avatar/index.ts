// packages/components-content/avatar
export { default as Avatar } from "./Avatar";
