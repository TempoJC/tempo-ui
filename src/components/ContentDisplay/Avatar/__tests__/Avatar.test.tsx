import { userEvent } from "@testing-library/user-event";
import { render, screen, waitFor } from "@testing-library/react";
import Avatar from "../Avatar";

describe("@tempo-ui/ContentDisplay/Avatar", () => {
  it("renders", () => {
    const { asFragment } = render(<Avatar initials="JG" imageUrl="http://some.avatar/url" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("adds the additional className", () => {
    const { container } = render(<Avatar initials="JG" className="customClass" />);

    expect(container.querySelector(".tempo-avatar")).toHaveClass("customClass");
  });

  it("renders in dark theme", () => {
    const { asFragment } = render(<Avatar initials="JG" color="dark" />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe("outlined type", () => {
    it("sets the outlined type", () => {
      const { container } = render(<Avatar initials="JG" type="outlined" />);

      expect(container.querySelector(".tempo-avatar--outlined")).toBeInTheDocument();
    });
  });

  describe("imageUrl", () => {
    describe("when provided", () => {
      it("renders an image", () => {
        const { container } = render(<Avatar initials="JG" imageUrl="http://some.avatar/url" />);

        const img = container.querySelector(".tempo-avatar__image");
        expect(img).toBeInTheDocument();
        expect(img).toHaveStyle({ backgroundImage: "url(http://some.avatar/url)" });
        expect(img).toHaveAttribute("title", "JG");
      });
    });

    describe("when not provided", () => {
      beforeEach(() => {
        jest.spyOn(global, "Image" as any).mockImplementation(() => {
          return {} as HTMLImageElement;
        });
      });

      afterEach(() => {
        jest.restoreAllMocks();
      });

      it("renders the initials and does not try to load the image", () => {
        render(<Avatar initials="JG" />);

        expect(screen.queryByText("JG")).toBeInTheDocument();

        expect(global.Image).not.toHaveBeenCalled();
      });
    });

    describe("when image fails to load", () => {
      let originalImageSrcSetter: PropertyDescriptor;

      beforeEach(() => {
        const descriptor = Object.getOwnPropertyDescriptor(global.Image.prototype, "src");
        if (descriptor) {
          originalImageSrcSetter = descriptor;
        } else {
          originalImageSrcSetter = {} as PropertyDescriptor; // Fallback
        }

        Object.defineProperty(global.Image.prototype, "src", {
          set(this: HTMLImageElement, v: string) {
            setTimeout(() => this.onerror?.(new Event("error")));
          },
        });
      });

      afterEach(() => {
        Object.defineProperty(global.Image.prototype, "src", originalImageSrcSetter);
      });

      it("renders initials", async () => {
        render(<Avatar imageUrl="invalid-image-url.png" initials="JG" testId="avatar" />);

        const initialsElement = await waitFor(() => screen.getByTestId("avatar-initials"));

        expect(initialsElement).toHaveTextContent("JG");
        expect(screen.queryByTestId("avatar-image")).not.toBeInTheDocument();
      });

      it("handles URL changes", async () => {
        const originalImageSrcSetter = Object.getOwnPropertyDescriptor(global.Image.prototype, "src");
        if (!originalImageSrcSetter) {
          throw new Error("No se pudo obtener el descriptor de la propiedad 'src' de Image.prototype");
        }

        let resolveOnLoad: ((value: unknown) => void) | undefined;
        const loadPromise = new Promise((res) => {
          resolveOnLoad = res;
        });

        let errorFired = false;

        Object.defineProperty(global.Image.prototype, "src", {
          set(this: HTMLImageElement, v: string) {
            setTimeout(() => {
              if (v.startsWith("valid")) {
                this.onload?.(new Event("load"));
                resolveOnLoad?.(undefined);
              } else {
                loadPromise.then(() => {
                  this.onerror?.(new Event("error"));
                  errorFired = true;
                });
              }
            });
          },
        });

        // initial render with an invalid image
        const { rerender } = render(<Avatar imageUrl="invalid-image-url.png" initials="JG" testId="avatar" />);

        // rerender with a valid image
        rerender(<Avatar imageUrl="valid-image-url.png" initials="JG" testId="avatar" />);

        const avatarImage = await screen.findByTestId("avatar-image");
        expect(avatarImage).toBeInTheDocument();

        // Wait until the error has been fired, which would
        // render the initials if the logic is not working
        await waitFor(() => expect(errorFired).toBe(true));

        expect(screen.queryByTestId("avatar-initials")).not.toBeInTheDocument();

        Object.defineProperty(global.Image.prototype, "src", originalImageSrcSetter);
      });
    });
  });

  describe("size is large", () => {
    it("renders in large size with image", () => {
      const { asFragment } = render(
        <Avatar initials="JG" color="dark" size="large" imageUrl="http://some.avatar/url" />,
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("renders in large size with initials", () => {
      const { asFragment } = render(<Avatar initials="JG" color="dark" size="large" />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("size is medium", () => {
    it("renders in medium size with image", () => {
      const { asFragment } = render(
        <Avatar initials="JG" color="dark" size="medium" imageUrl="http://some.avatar/url" />,
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("renders in medium size with initials", () => {
      const { asFragment } = render(<Avatar initials="JG" color="dark" size="medium" />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("when onClick is set", () => {
    it("is executed when the Avatar is clicked", async () => {
      const clickHandler = jest.fn();
      const { container } = render(<Avatar initials="JG" onClick={clickHandler} />);

      const avatar = container.querySelector(".tempo-avatar");
      if (avatar) {
        await userEvent.click(avatar);
      } else {
        throw new Error("Avatar element not found.");
      }

      expect(clickHandler).toHaveBeenCalledTimes(1);
    });

    it("adds a special className to the component", () => {
      const { container } = render(<Avatar initials="JG" onClick={() => {}} />);

      expect(container.querySelector(".tempo-avatar--clickable")).toBeInTheDocument();
    });
  });

  describe("testId prop", () => {
    it("renders the avatar with the specified data-testid attribute", () => {
      const { getByTestId } = render(<Avatar initials="JG" testId="avatar" />);
      const element = getByTestId("avatar");
      expect(element).toBeInTheDocument();
      expect(element).toHaveAttribute("data-testid", "avatar");
    });
  });
});
