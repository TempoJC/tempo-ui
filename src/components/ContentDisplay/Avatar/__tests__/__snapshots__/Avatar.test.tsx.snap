// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`@tempo-ui/ContentDisplay/Avatar renders 1`] = `
<DocumentFragment>
  <div
    class="tempo-avatar"
  >
    <div
      class="tempo-avatar__image"
      style="background-image: url("http://some.avatar/url");"
      title="JG"
    />
  </div>
</DocumentFragment>
`;

exports[`@tempo-ui/ContentDisplay/Avatar renders in dark theme 1`] = `
<DocumentFragment>
  <div
    class="tempo-avatar tempo-avatar--color-dark"
  >
    <div
      class="tempo-avatar__initials"
    >
      JG
    </div>
  </div>
</DocumentFragment>
`;

exports[`@tempo-ui/ContentDisplay/Avatar size is large renders in large size with image 1`] = `
<DocumentFragment>
  <div
    class="tempo-avatar tempo-avatar--large tempo-avatar--color-dark"
  >
    <div
      class="tempo-avatar__image"
      style="background-image: url("http://some.avatar/url");"
      title="JG"
    />
  </div>
</DocumentFragment>
`;

exports[`@tempo-ui/ContentDisplay/Avatar size is large renders in large size with initials 1`] = `
<DocumentFragment>
  <div
    class="tempo-avatar tempo-avatar--large tempo-avatar--color-dark"
  >
    <div
      class="tempo-avatar__initials"
    >
      JG
    </div>
  </div>
</DocumentFragment>
`;

exports[`@tempo-ui/ContentDisplay/Avatar size is medium renders in medium size with image 1`] = `
<DocumentFragment>
  <div
    class="tempo-avatar tempo-avatar--medium tempo-avatar--color-dark"
  >
    <div
      class="tempo-avatar__image"
      style="background-image: url("http://some.avatar/url");"
      title="JG"
    />
  </div>
</DocumentFragment>
`;

exports[`@tempo-ui/ContentDisplay/Avatar size is medium renders in medium size with initials 1`] = `
<DocumentFragment>
  <div
    class="tempo-avatar tempo-avatar--medium tempo-avatar--color-dark"
  >
    <div
      class="tempo-avatar__initials"
    >
      JG
    </div>
  </div>
</DocumentFragment>
`;
